File: programmers/Level2/게임맵최단거리.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    int[] dy = {1, 0, -1, 0};
    int[] dx = {0, 1, 0, -1};
    
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        boolean[][] visited = new boolean[n][m];
        
        int answer = bfs(n, m, maps, visited);
        return answer;
    }
    
    int bfs(int n, int m, int[][] maps, boolean[][] visited) {
        Queue<Pos> q = new LinkedList<>();
        q.offer(new Pos(0, 0, 1));
        visited[0][0] = true;
        
        while (!q.isEmpty()) {
            Pos p = q.poll();
            
            if (p.y == n-1 && p.x == m-1) {
                return p.count;
            }
            
            for (int i = 0; i < 4; i++) {
                int ny = p.y + dy[i];
                int nx = p.x + dx[i];
                
                if (ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
                if (visited[ny][nx] || maps[ny][nx] == 0) continue;
                
                visited[ny][nx] = true;
                q.offer(new Pos(ny, nx, p.count + 1));
            }
        }
        
        return -1;
    }
}

class Pos {
    int y, x;
    int count;
    
    Pos(int y, int x, int count) {
        this.y = y;
        this.x = x;
        this.count = count;
    }
}

File: programmers/Level2/게임맵최단거리.py
from collections import deque

def solution(maps):
    dx = [1, 0, -1, 0]
    dy = [0, 1, 0, -1]
    
    n, m = len(maps), len(maps[0])
    
    queue = deque()
    queue.append([0, 0])
    
    graph = [[-1 for _ in range(m)] for _ in range(n)]
    graph[0][0] = 1
    
    while queue:
        y, x = queue.popleft()
        
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            
            if 0 <= ny < n and 0 <= nx < m and maps[ny][nx] == 1:
                if graph[ny][nx] == -1:
                    graph[ny][nx] = graph[y][x] + 1
                    queue.append([ny, nx])
    
    return graph[-1][-1]